sdk/core/allocator/alloc.h

```diff
@@ -489,24 +757,127 @@ class __packed __aligned(MALLOC_ALIGNMENT) TChunk : public MChunk
 	TChunk *parent;
 	// the tree index this chunk is in
 	BIndex index;
+	/*
+	 * There's padding to alignment here, so if we needed more metadata, it
+	 * could be "free", in spatial terms.
+	 *
+	 * This padding doesn't break our targeted zeroing in metadata_clear()
+	 * because said padding will be zeroed before we view the chunk as a TChunk
+	 * and we won't store to it.
+	 */
+
+	TChunk *leftmost_child()
+	{
+		if (child[0] != nullptr)
+			return child[0];
+		return child[1];
+	}
+
+	static constexpr uintptr_t RingParent = 0;
+	static constexpr uintptr_t RootParent = 1;
+
+	bool is_root()
+	{
+		return reinterpret_cast<uintptr_t>(parent) == RootParent;
+	}
+	void mark_root()
+	{
+		parent = reinterpret_cast<TChunk *>(RootParent);
+	}
+
+	bool is_tree_ring()
+	{
+		return parent == reinterpret_cast<TChunk *>(RingParent);
+	}
+	void mark_tree_ring()
```

Comment:
Nit: could benefit from more comments here (although the comments before this PR were already lacking). I think this is called for tree chunks that are not the direct child of a parent?